src/pdf_parser/pdf_utils/ocr.py

```diff
@@ -0,0 +1,487 @@
+    text_blocks_to_keep = []
+    for block in block_text_segments:
+        if block in paragraph_text_segments:
+            continue
+        else:
+            text_blocks_to_keep.append(block)
+
+    combined_text_segments = text_blocks_to_keep + paragraph_text_segments
+
+    return (
+        fully_structured_blocks,
+        combined_text_segments,
+        block_text_segments,
+        paragraph_text_segments,
+    )
+
+
+def google_coords_to_lp_coords(
+    google_coords: RepeatedCompositeContainer,
+) -> Tuple[int, int, int, int]:
+    """Converts Google OCR coordinates to LayoutParser coordinates.
+
+    Args:
+        google_coords: Google OCR coordinates.
+
+    Returns:
+        Tuple of (x1, y1, x2, y2)
+    """
+    x1, y1, x2, y2 = (
+        google_coords[0].x,
+        google_coords[0].y,
+        google_coords[2].x,
+        google_coords[2].y,
+    )
+    return x1, y1, x2, y2
+
+
+def combine_google_lp(
+    image,
+    google_layout: List[GoogleTextSegment],
+    lp_layout: Layout,
+    threshold: float = 0.9,
+    top_exclude: float = 0.1,
+    bottom_exclude: float = 0.1,
+):
+    """
+    Combine google layout with layoutparser layout.
+
+    - Replaces layoutparser objects with objects recognised by google API + their text if they overlap sufficiently.
+    - Does not include google objects if they seem to be not part of the main text (e.g. headers, footers, etc.). We
+    use top exclude and bottom exclude to ascertain this.
+
+
+    Args:
+        image: The image the layout is based on.
+        google_layout: List of GoogleTextSegment objects inferred from the structure returned by the Google OCR API.
+        lp_layout: Layout object
+        threshold: Threshold for overlap between google and layoutparser objects. If the overlap is larger than
+            threshold, the layoutparser object is replaced by the google object.
+        top_exclude: Exclude objects from the top of the page if they are above this fraction of the page.
+        bottom_exclude: Exclude objects from the bottom of the page if they are below this fraction of the page.
+
+    Returns:
+        Layout object with google objects replacing layoutparser objects if they overlap sufficiently.
+    """
+    shapely_google = [google_vertex_to_shapely(b.coordinates) for b in google_layout]
+    shapely_layout = [lp_coords_to_shapely_polygon(b.coordinates) for b in lp_layout]
+    # for every block in the google layout, find the fraction of the block that is covered by the layoutparser layout
+    dd_intersection_over_union = defaultdict(list)
+    for ix_goog, google_block in enumerate(shapely_google):
+        for ix_lp, lp_block in enumerate(shapely_layout):
+            # Find the intersection over union of the two blocks.
+            intersection = google_block.intersection(lp_block).area
+            union = google_block.union(lp_block).area
+            dd_intersection_over_union[ix_goog].append(intersection / union)
+
```

Comment:
In testing on a few files this works really well. Google picks up the same blocks as LayoutParser + heuristics for the most part. There are definitely more improvements we can leverage from google's document API too, but this is a minimal working version.